### PR TITLE
fix(runtime_deps): check paths are descendant of the library root

### PR DESCRIPTION
**Pull request by anmonteiro**
- add a new error when `melange.runtime_deps` in public libraries refer to paths outside of the library src dir

### CODE REVIEW

File: src/dune_rules/install_rules.ml [hunk 1]
```diff
@@ -43,6 +43,27 @@ module Package_paths = struct
     Path.Build.extend_basename (meta_file ctx pkg) ~suffix:".template"
 end
 
+let check_runtime_deps_relative_path path ~loc ~lib_info =
+  let lib_src_dir = Lib_info.src_dir lib_info in
+  match
+    Path.Local.descendant
+      (path |> Path.as_in_build_dir_exn |> Path.Build.local)
```
Comment (rgrinberg) [inline src/dune_rules/install_rules.ml:50]: How about have the caller do this cast instead?

File: test/blackbox-tests/test-cases/melange/installed-library-with-runtime-deps-nested-errors.t [hunk 2]
```diff
@@ -0,0 +1,35 @@
+Test errors when public library runtime dependencies escape the dune file dir
```
Comment (rgrinberg) [inline test/blackbox-tests/test-cases/melange/installed-library-with-runtime-deps-nested-errors.t:1]: Can the test come be a separate commit?

File: src/dune_rules/install_rules.ml (hunk 1 above)
Comment (anmonteiro) [reply src/dune_rules/install_rules.ml:50]: How would you suggest getting back the original path to return from this function? (currently line 65)

Alternative I can think of:
- make this function return `unit` and just check the path?

File: test/blackbox-tests/test-cases/melange/installed-library-with-runtime-deps-nested-errors.t (hunk 2 above)
Comment (anmonteiro) [reply test/blackbox-tests/test-cases/melange/installed-library-with-runtime-deps-nested-errors.t:1]: did this.

File: src/dune_rules/install_rules.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_rules/install_rules.ml:50]: > make this function return unit and just check the path?

I think that would be even better.

Comment (anmonteiro) [reply src/dune_rules/install_rules.ml]: Ok I pushed that alternative.

File: src/dune_rules/install_rules.ml [hunk 3]
```diff
@@ -43,6 +43,25 @@ module Package_paths = struct
...
 
+let check_runtime_deps_relative_path local_path ~loc ~lib_info =
+  let lib_src_dir = Lib_info.src_dir lib_info in
+  match
+    Path.Local.descendant local_path ~of_:(Path.Build.local lib_src_dir)
+  with
+  | None ->
+    User_error.raise ~loc
+      [ Pp.textf
+          "Public library %s depends on a path defined in a parent directory. \
```
Comment (jchavarri) [inline src/dune_rules/install_rules.ml:54]: Not sure if "in a parent directory" could be refined, because afaiu it doesn't necessarily need to be a parent directory.

Alternative: "Public library %s depends on an asset that is outside the library directory tree."

**Review by jchavarri — APPROVED** [review]
Looks great! Afaiu this constraints a bit more the way runtime deps work, right? From being able to depend on any runtime asset in a project, to be able to depend on any runtime asset in a library? I wonder if the docs should be updated to reflect this new constraint.

**Comment by anmonteiro** [comment]
> this constraints a bit more the way runtime deps work, right?

the only additional constraint this imposes is that _public_ libraries need to have runtime_deps under their "lib root" (the folder where the dune file is).

File: src/dune_rules/install_rules.ml [hunk 4]
```diff
@@ -178,7 +197,13 @@ end = struct
           ~paths:(Disallow_external lib_name)
       in
       Path.Set.to_list_map deps ~f:(fun path ->
-          let path = Path.as_in_build_dir_exn path in
+          let path =
+            let local_path =
+              path |> Path.as_in_build_dir_exn |> Path.Build.local
+            in
+            check_runtime_deps_relative_path ~lib_info:info ~loc local_path;
+            Path.as_in_build_dir_exn path
```
Comment (rgrinberg) [inline src/dune_rules/install_rules.ml:205]: I think calling `Path.as_in_build_dir_exn` should only be called once.

Comment (anmonteiro) [reply src/dune_rules/install_rules.ml]: Good point, just fixed that.

**Review by rgrinberg — APPROVED** [review]

File: src/dune_rules/install_rules.ml (hunk 3 above)
Comment (anmonteiro) [reply src/dune_rules/install_rules.ml]: Agree. Tweaked the message based on this suggestion.
